phoenix-core/src/main/java/org/apache/phoenix/query/ConnectionQueryServicesImpl.java

```diff
@@ -3101,49 +3124,68 @@ void ensureSystemTablesUpgraded(ReadOnlyProps props)
+            // If we cannot acquire lock, it means some old client is either migrating SYSCAT or trying to upgrade the
+            // schema of SYSCAT table and hence it should not be interrupted
+            acquiredMutexLock = acquireUpgradeMutex(0, mutexRowKey);
+            if(acquiredMutexLock) {
+                logger.debug("Acquired lock in SYSMUTEX table for migrating SYSTEM tables to SYSTEM namespace");
+            }
+            // We will not reach here if we fail to acquire the lock, since it throws UpgradeInProgressException
+
+            // Handle the upgrade of SYSMUTEX table separately since it doesn't have any entries in SYSCAT
+            String sysMutexSrcTableName = PhoenixDatabaseMetaData.SYSTEM_MUTEX_NAME;
+            String sysMutexDestTableName = SchemaUtil.getPhysicalName(sysMutexSrcTableName.getBytes(), props).getNameAsString();
+            UpgradeUtil.mapTableToNamespace(admin, sysMutexSrcTableName, sysMutexDestTableName, PTableType.SYSTEM);
+
             byte[] mappedSystemTable = SchemaUtil
                     .getPhysicalName(PhoenixDatabaseMetaData.SYSTEM_CATALOG_NAME_BYTES, props).getName();
             metatable = getTable(mappedSystemTable);
             if (tableNames.contains(PhoenixDatabaseMetaData.SYSTEM_CATALOG_HBASE_TABLE_NAME)) {
                 if (!admin.tableExists(mappedSystemTable)) {
+                    // Actual migration of SYSCAT table
                     UpgradeUtil.mapTableToNamespace(admin, metatable,
                             PhoenixDatabaseMetaData.SYSTEM_CATALOG_NAME, props, null, PTableType.SYSTEM,
                             null);
+                    // Invalidate the client-side metadataCache
                     ConnectionQueryServicesImpl.this.removeTable(null,
                             PhoenixDatabaseMetaData.SYSTEM_CATALOG_NAME, null,
                             MetaDataProtocol.MIN_SYSTEM_TABLE_TIMESTAMP_4_1_0);
                 }
                 tableNames.remove(PhoenixDatabaseMetaData.SYSTEM_CATALOG_HBASE_TABLE_NAME);
             }
-            tableNames.remove(PhoenixDatabaseMetaData.SYSTEM_MUTEX_HBASE_TABLE_NAME);
             for (TableName table : tableNames) {
                 UpgradeUtil.mapTableToNamespace(admin, metatable, table.getNameAsString(), props, null, PTableType.SYSTEM,
                         null);
                 ConnectionQueryServicesImpl.this.removeTable(null, table.getNameAsString(), null,
                         MetaDataProtocol.MIN_SYSTEM_TABLE_TIMESTAMP_4_1_0);
             }
-            if (!tableNames.isEmpty()) {
-                clearCache();
-            }
+
+            // Clear the server-side metadataCache when all tables are migrated so that the new PTable can be loaded with NS mapping
+            clearCache();
         } finally {
             if (metatable != null) {
                 metatable.close();
             }
+            if(acquiredMutexLock) {
```

Comment:
Why wasn't this  "UpgradeUtil.mapTableToNamespace" synchronized in the existing code?

Reply:
@joshelser @ankitsinghal Can probably comment on this one.

Reply:
I have no good explanation for why it was the way it was when I last looked at it. I can only assumed I left it as it was (really, just extracted part of this one method into another). If you think it should also be marked synchronized, go for it.

Reply:
According to my understanding and discussions with @twdsilva, I believe that multiple clients should not be allowed to carry on this portion of code simultaneously. For example, a client is trying to upgrade the SYSCAT version and another client connects with the property `phoenix.schema.mapSystemTablesToNamespace` to true. The latter client might disable SYSCAT and enable it to SYS:CAT, in such a case the client operations from former client might leave SYSCAT inconsistent.
@joshelser Let me know your thoughts about it.

Reply:
If there isn't any mutual exclusion happening at a higher level, most definitely we should make sure this isn't happening concurrently.

What about clients in separate JVMs though? What's the "prior art" for the rest of the SYSTEM tables?

Reply:
This SYSMUTEX table is in place to make sure that clients in separate JVM's don't carry out the same task at the same time. I think @twdsilva and @samarthjain jain discovered the bug where multiple clients tried to upgrade SYSCAT at the same time and that is when the SYSMUTEX table was introduced. I feel that a similar race condition can be triggered over here, so a mutual exclusion is required, which is done via SYSMUTEX table. 
Does this answer your question?

Reply:
Yup, fine by me. It seems to me to be the same as it was previously :) The mutex table update would provide exclusion for both same and different JVM cases.

Reply:
Earlier, we didn't have any locking during table migration. So its a race condition which can lead to unexpected results.